### PR TITLE
8332176: Refactor ClassListParser::parse()

### PR DESCRIPTION
**Pull request by iklam**
The handling of `@` tags in `ClassListParser::parse()` is quite convoluted. As we expect to add more `@` tags in Project Leyden, we should refactor this code to make it more maintainable.

It's easier to review the changes by ignoring the whitespace changes.

- I refactored the parsing to two parts -- `parse_at_tags()` and `parse_class_name_and_attributes()`
- I changed  the `ClassListParser::parse()` to no longer return the number of classes loaded -- some `@` tags can cause many classes to be loaded as a side effect (e.g., `@lambda-form-invoker`), so the number returned by `ClassListParser::parse()` is not accurate. There are other CDS logs that print out the exact number of classes stored in the archive.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8332176](https://bugs.openjdk.org/browse/JDK-8332176): Refactor ClassListParser::parse() (**Enhancement** - P4)

### Reviewers
 * [Matias Saavedra Silva](https://openjdk.org/census#matsaave) (@matias9927 - **Reviewer**)
 * [Calvin Cheung](https://openjdk.org/census#ccheung) (@calvinccheung - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19225/head:pull/19225` \
`$ git checkout pull/19225`

Update a local copy of the PR: \
`$ git checkout pull/19225` \
`$ git pull https://git.openjdk.org/jdk.git pull/19225/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19225`

View PR using the GUI difftool: \
`$ git pr show -t 19225`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19225.diff">https://git.openjdk.org/jdk/pull/19225.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19225#issuecomment-2109124794)

### CODE REVIEW

**Review by matias9927 — COMMENTED** [review]
Thanks for the refactor, this is a bit easier to read now! I have a couple of considerations below:

File: src/hotspot/share/cds/classListParser.hpp [hunk 1]
```diff
@@ -129,9 +130,9 @@ class ClassListParser : public StackObj {
   ~ClassListParser();
 
 public:
-  static int parse_classlist(const char* classlist_path, ParseMode parse_mode, TRAPS) {
+  static void parse_classlist(const char* classlist_path, ParseMode parse_mode, TRAPS) {
     ClassListParser parser(classlist_path, parse_mode);
-    return parser.parse(THREAD); // returns the number of classes loaded.
+    parser.parse(THREAD); // returns the number of classes loaded.
```
Comment (calvinccheung) [inline src/hotspot/share/cds/classListParser.hpp:135]: The comment is no longer applicable.

**Review by calvinccheung — COMMENTED** [review]
Looks good overall. Found an outdated comment.

File: src/hotspot/share/cds/classListParser.cpp [hunk 2]
```diff
@@ -257,10 +245,9 @@ bool ClassListParser::parse_one_line() {
   //     # the class is loaded from classpath
   //     id may be specified
   //     super, interfaces, loader must not be specified
-  return true;
 }
 
-void ClassListParser::split_tokens_by_whitespace(int offset) {
+void ClassListParser::split_tokens_by_whitespace(int offset, GrowableArray<const char*>* items) {
```
[the remark this answers is not in the document]
Comment (iklam) [reply src/hotspot/share/cds/classListParser.cpp:250]: I changed this function as its name is confusing. If we want to keep the existing code that operates on `_indy_items`, we have to rename it accordingly, as `split_tokens_by_whitespace` doesn't mention anything about  `_indy_items`. But that seems to be going the wrong direction, making it more difficult to reuse this function in the future, if such need arises.

I think it's a good idea to reduce coupling in the code. This function splits a line into multiple parts, so we shouldn't make it depend on `_indy_items`.

**Review by calvinccheung — APPROVED** [review]

**Review by matias9927 — APPROVED** [review]
Updates look good, thanks!

**Comment by iklam** [comment]
Thanks @calvinccheung and @matias9927 for the review
/integrate

**Pull request closed without merging**